### PR TITLE
[ADP-2565] Clean up package `delta-table`

### PR DESCRIPTION
**Pull request by HeinrichApfelmus**
This pull request cleans up the `delta-table` package in order to prepare the implementation of a library for working with SQL tables.

Specifically, this pull requests

* Splits off existing code into a `delta-chain` package, to be cleaned up later.
* Removes most modules from the `delta-table` package.

### Issue Number

ADP-2565

### CODE REVIEW

File: lib/delta-chain/README.md [hunk 1]
```diff
@@ -0,0 +1,22 @@
+## Overview
+
+At the time of this writing, the package contains
+an early prototype of a `Table` type that represents
+database tables, and a use case in `Data.Chain`.
```
Comment (paweljakubas) [inline lib/delta-chain/README.md]:
```suggestion
database tables, and an example usage can be found in `Data.Chain`.
```

File: lib/delta-chain/src/Data/Table.hs [hunk 2]
```diff
@@ -0,0 +1,298 @@
...
+    Table
+-------------------------------------------------------------------------------}
+
+-- | A 'Table' is a collection of rows.
+data Table row = Table
+    { rows :: IntMap row
+    -- ^ Rows indexed by unique ID.
+    , uids :: Supply
+    -- ^ Unique ID supply.
+    -- WARNING: This is an internal part of the structure.
```
Comment (paweljakubas) [inline lib/delta-chain/src/Data/Table.hs:95]: if so what about not persuading users to construct `Table` like above but add smart constructors/pattern synonyms hiding `uids` ?

File: lib/delta-chain/src/Data/Table.hs [hunk 3]
```diff
@@ -0,0 +1,298 @@
...
+    (Pile x) == (Pile y) = sort x == sort y
+
+fromSet :: Set a -> Pile a
+fromSet = Pile . Set.toList
+
+-- | Randomly permute the objects in a 'Pile'. Useful for stress testing.
+--
+-- Every function @f :: Pile A -> B@ should satisfy
+--
+-- > forall g.  f . permute g = f
```
Comment (paweljakubas) [inline lib/delta-chain/src/Data/Table.hs:239]: :+1:

**Review by paweljakubas — APPROVED** [review]
nice stuff! I wonder if we should not add tests and properties handling functions in `Data.Table` . Like we start with `empty` and `insertMany` (here two partitions created by some predicate), then we `selectWhere` with some predicate and always get one partition...etc.  Also wonder if instead list we should not use Vector or array as a way of storing. I am even more persuaded we should diligently test `tableIntoDatabase`. Of course could be done in the next PRs

File: lib/delta-chain/src/Data/Table.hs (hunk 2 above)
Comment (HeinrichApfelmus) [reply lib/delta-chain/src/Data/Table.hs:95]: Yes. At the moment, the module `Data.Table` is preliminary, and will be adjusted later. I think that I had exposed the internal constructor in order to implement reading / writing `Data.Table` from an SQLite database table.

**Comment by HeinrichApfelmus** [comment]
> I wonder if we should not add tests and properties handling functions in Data.Table .

Yes, at a later time. At the moment, the module `Data.Table` is preliminary. I will add a comment to that effect.

File: lib/delta-chain/README.md (hunk 1 above)
Comment (HeinrichApfelmus) [reply lib/delta-chain/README.md]: done

**Pull request merged**